File: test_op/ascend-transformer-boost/tests/unittest/kernels/elewise/test_dynamic_quant.cpp
/*
 * Copyright (c) 2024 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */
#include <ATen/ATen.h>
#include <gtest/gtest.h>
#include <iostream>
#include <cmath>
#include <mki/utils/fp16/fp16_t.h>
#include <mki/utils/log/log.h>
#include "asdops/params/elewise.h"
#include "device_version_check.h"
#include "test_utils/float_util.h"
#include "test_utils/op_test.h"
#include "test_utils/golden.h"
#include "op_desc_json.h"

using namespace AsdOps;
using namespace Mki;
static constexpr float HALF_FLOAT_MIN = -1;
static constexpr float HALF_FLOAT_MAX = 1;
static constexpr float ATOL = 0.001;
static constexpr float RTOL = 0.001;
static constexpr float QUANT_MAX_VALUE = 127;
static constexpr float QUANT_ASYMMETRIC_VALUE = 255;

namespace AsdOps {
Status DQMatchAtTensorFloat(float atol, float rtol, at::Tensor &out, at::Tensor &gt)
{
    float *outArr = (float *)out.storage().data_ptr().get();
    float *gtArr = (float *)gt.storage().data_ptr().get();
    for (int i = 0; i < out.numel(); i++) {
        float expect = gtArr[i];
        float actual = outArr[i];
        if (!Mki::Test::FloatUtil::FloatJudgeEqual(expect, actual, ATOL, RTOL)) {
            return Status::FailStatus(-1, "float judge not equal");
        }
    }
    return Status::OkStatus();
}

Status DQMatchAtTensorInt(float atol, float rtol, at::Tensor &out, at::Tensor &gt)
{
    int64_t countRight = 0;
    int64_t countFalse = 0;
    int8_t *result0 = static_cast<int8_t *>(out.storage().data_ptr().get());
    float *expect0 = static_cast<float *>(gt.storage().data_ptr().get());
    for (int i = 0; i < out.numel(); i++) {
        int actual = result0[i];
        int expect = std::round(expect0[i]);
        if (expect >= 128) {
            expect = 127;
        }
        if (expect != actual) {
            if (abs(abs(expect) - abs(actual)) == 1) {
                countFalse = countFalse + 1;
            }
        } else {
            countRight = countRight + 1;
        }
    }
    MKI_LOG(INFO) << "out.numel(): " << out.numel() << "\tcountRight: " << countRight \
        << "\tcountFalse: " << countFalse << "\tcorrectRate: " << (float)countRight / out.numel();
    if (!Mki::Test::FloatUtil::FloatJudgeEqual(float(out.numel()), float(countRight + countFalse), atol, rtol)) {
        std::string msg =
            "expect: " + std::to_string(out.numel()) + ", result: " + std::to_string(countRight + countFalse);
        return Status::FailStatus(-1, msg);
    }
    return Status::OkStatus();
}

Status DynamicQuantGolden(float atol, float rtol, OpParam::Elewise opParam, const Mki::Test::GoldenContext &context)
{
    const Tensor &inputX = context.hostInTensors.at(0);
    const Tensor &tensorOut = context.hostOutTensors.at(0);
    const Tensor &tensorOutScale = context.hostOutTensors.at(1);
    const Tensor &tensorOutOffset = context.hostOutTensors.at(2);
    std::vector<int64_t> shape = {inputX.desc.dims[0], inputX.desc.dims[1], inputX.desc.dims[2]};
    std::vector<int64_t> scaleShape = {inputX.desc.dims[0], inputX.desc.dims[1]};
    auto attensorOut0 = at::from_blob(tensorOut.data, shape, at::kChar);
    auto attensorOut1 = at::from_blob(tensorOutScale.data, scaleShape, at::kFloat);
    auto attensorOut2 = at::from_blob(tensorOutOffset.data, scaleShape, at::kFloat);
    auto attensorInputX = at::from_blob(inputX.data, shape, at::kHalf);
    Status status;
    int64_t dim = inputX.desc.dims.size() - 1; // unsqueeze the last dim
    if (opParam.asymmetric) {
        auto resMax = std::get<0>(at::max(attensorInputX, dim, true));
        auto resMin = std::get<0>(at::min(attensorInputX, dim, true));
        auto scale = at::sub(resMax, resMin);
        auto offset = at::add(resMax, resMin);
        offset = at::div(offset, scale).to(at::kFloat);
        scale = at::div(scale, QUANT_ASYMMETRIC_VALUE).to(at::kFloat);
        offset = at::mul(offset, -127.5f); // -127.5 is -255 / 2
        auto outputTensor = at::div(attensorInputX, scale).to(at::kFloat);
        outputTensor = at::add(outputTensor, offset);
        auto groundtruth1 = at::unsqueeze(scale, dim);
        auto groundtruth2 = at::unsqueeze(offset, dim);
        MKI_LOG(INFO) << "check offset value";
        status = DQMatchAtTensorFloat(atol, rtol, attensorOut2, groundtruth2);
        if (!status.Ok()) {
            return status;
        }
        MKI_LOG(INFO) << "check scale value";
        status = DQMatchAtTensorFloat(atol, rtol, attensorOut1, groundtruth1);
        if (!status.Ok()) {
            return status;
        }
        return DQMatchAtTensorInt(atol, rtol, attensorOut0, outputTensor);
    } else {
        auto resAbs = at::abs(attensorInputX);
        auto resMax = std::get<0>(at::max(resAbs, dim, true));
        auto outputTensor = at::div(attensorInputX, resMax).to(at::kFloat);
        outputTensor = at::mul(outputTensor, QUANT_MAX_VALUE);
        resMax = at::div(resMax, QUANT_MAX_VALUE).to(at::kFloat);
        auto groundtruth1 = at::unsqueeze(resMax, dim);
        MKI_LOG(INFO) << "check scale value";
        status = DQMatchAtTensorFloat(atol, rtol, attensorOut1, groundtruth1);
        if (!status.Ok()) {
            return status;
        }
        return DQMatchAtTensorInt(atol, rtol, attensorOut0, outputTensor);
    }
}
} // namespace AsdOps

TEST(DynamicQuantFP16, DynamicQuantFP16Case1)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 1;
    uint32_t b = 70;
    uint32_t c = 32;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case2)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 64;
    uint32_t b = 1;
    uint32_t c = 32;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case3)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 1;
    uint32_t b = 1;
    uint32_t c = 32;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case4)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 1;
    uint32_t b = 1;
    uint32_t c = 64;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case5)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 1;
    uint32_t b = 1;
    uint32_t c = 128;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case6)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 6144;
    uint32_t b = 1;
    uint32_t c = 4096;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case7)
{
    CHECK_DEVICE_VERSION_NOT_ASCEND910A();
    CHECK_DEVICE_VERSION_NOT_ASCEND310B();
    uint32_t a = 123;
    uint32_t b = 1;
    uint32_t c = 4096;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case8)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 1;
    uint32_t b = 50;
    uint32_t c = 50;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case9)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 1;
    uint32_t b = 1;
    uint32_t c = 100;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case10)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 1;
    uint32_t b = 1;
    uint32_t c = 300;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case11)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 123;
    uint32_t b = 1;
    uint32_t c = 1000;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case12)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 5000;
    uint32_t b = 1;
    uint32_t c = 4100;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = true;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

TEST(DynamicQuantFP16, DynamicQuantFP16Case13)
{
    CHECK_DEVICE_VERSION_ASCEND910B();
    uint32_t a = 5000;
    uint32_t b = 1;
    uint32_t c = 4100;
    Mki::Test::MkiOpTest opTest;
    AsdOps::OpParam::Elewise opParam;
    Mki::Test::UtOpDesc opDesc;
    opDesc.opName = "ElewiseOperation";
    opParam.elewiseType = AsdOps::OpParam::Elewise::ELEWISE_DYNAMIC_QUANT;
    opParam.asymmetric = false;
    opDesc.specificParam = opParam;

    opTest.Golden(std::bind(DynamicQuantGolden, ATOL, RTOL, opParam, std::placeholders::_1));
    opTest.FloatRand(HALF_FLOAT_MIN, HALF_FLOAT_MAX);
    SVector<TensorDesc> inTensorDesc = {{TENSOR_DTYPE_FLOAT16, TENSOR_FORMAT_ND, {a, b, c}}};
    Status status = opTest.Run(opDesc, inTensorDesc);
    ASSERT_EQ(status.Ok(), true);
}

File: test_op/ascend-transformer-boost/tests/apitest/kernelstest/elewise/test_dynamic_quant.py
#
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
#
import unittest
import numpy as np
import torch
import op_test


OP_NAME = "ElewiseOperation"
PARAM = {"elewiseType" : 20, "asymmetric" : False}


class TestDynamicQuant(op_test.OpTest):

    def golden_calc(self, in_tensors):
        input_x = in_tensors[0].to(torch.float).cpu().numpy()
        if PARAM["asymmetric"]:
            row_max = np.max(input_x, axis=-1, keepdims=True)
            row_min = np.min(input_x, axis=-1, keepdims=True)
            row_max = row_max.astype(np.float32)
            row_min = row_min.astype(np.float32)
            out_scale = (row_max - row_min) / 255
            out_offset = - (row_max + row_min) / (2 * out_scale)

            input_x = input_x.astype(np.float32)
            input_x = input_x / out_scale
            input_x = input_x + out_offset
            input_x = np.clip(input_x, -128, 127)
            out_x = np.round(input_x)
            return [torch.from_numpy(out_x).to(torch.int8),
                    torch.from_numpy(out_scale.squeeze(axis=-1)).to(torch.float32),
                    torch.from_numpy(out_offset.squeeze(axis=-1)).to(torch.float32)]
        else:
            input_abs = np.abs(input_x)
            scale = np.max(input_abs, axis=-1, keepdims=True)
            scale = scale.astype(np.float32)
            out_scale = scale / 127
            input_x = input_x * 127
            input_x = input_x / scale
            out_x = np.round(input_x)
            return [torch.from_numpy(out_x).to(torch.int8),
                    torch.from_numpy(out_scale.squeeze(axis=-1)).to(torch.float32)]

    def golden_compare(self, out_tensor, golden_out_tensor):
        
        diff = torch.abs(torch.subtract(out_tensor[0], golden_out_tensor[0]))
        if torch.any(torch.greater(diff, 1)):
            print("[new standards] output0 accuracy failed")
            return False

        diff = torch.abs(torch.subtract(out_tensor[1], golden_out_tensor[1]))
        tensor_max = torch.maximum(torch.ones(golden_out_tensor[1].shape, dtype=golden_out_tensor[1].dtype),
                                   torch.abs(golden_out_tensor[1]))

        if torch.any(torch.greater(diff, 2**(-8) * tensor_max)):
            print("[new standards] output1 accuracy failed")
            return False

        if torch.any(torch.greater(torch.abs(torch.mean(torch.div(diff, tensor_max))), 2**(-10))):
            print("[new standards] output1 eb failed")
            return False

        if len(golden_out_tensor) > 2:
            diff = torch.abs(torch.subtract(out_tensor[2], golden_out_tensor[2]))
            tensor_max = torch.maximum(torch.ones(golden_out_tensor[2].shape, dtype=golden_out_tensor[1].dtype),
                                    torch.abs(golden_out_tensor[2]))

            if torch.any(torch.greater(diff, 2**(-8) * tensor_max)):
                print("[new standards] output2 accuracy failed")
                return False

            if torch.any(torch.greater(torch.abs(torch.mean(torch.div(diff, tensor_max))), 2**(-10))):
                print("[new standards] output2 eb failed")
                return False
        return True

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case1(self):
        shape0 = (2, 32, 32)
        shape1 = (2, 32)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case2(self):
        shape0 = (2, 32, 32)
        shape1 = (2, 32)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = True
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case3(self):
        shape0 = (200, 3, 1024)
        shape1 = (200, 3)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case4(self):
        shape0 = (300, 2, 2048)
        shape1 = (300, 2)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = True
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case5(self):
        shape0 = (200, 1024)
        shape1 = (200)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case6(self):
        shape0 = (300, 2048)
        shape1 = (300)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = True
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    # when dim=4、5, pta bug, test case failed
    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case7(self):
        shape0 = (2, 5, 10, 3, 3, 1024)
        shape1 = (2, 5, 10, 3, 3)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])

    @op_test.skip_310b
    @op_test.skip_910a
    def test_dynamic_quant_fp16_case8(self):
        shape0 = (2, 5, 2, 5, 3, 2, 2048)
        shape1 = (2, 5, 2, 5, 3, 2)
        input0 = np.random.uniform(low=-5, high=10, size=shape0).astype(np.float16)
        PARAM["asymmetric"] = True
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0)],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])
        
    @op_test.only_910b
    def test_dynamic_quant_bf16_case9(self):
        shape0 = (3072, 5120)
        shape1 = (3072)
        input0 = np.random.uniform(low=-5, high=5, size=shape0).astype(np.float32)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0).bfloat16()],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])
        
    @op_test.only_910b
    def test_dynamic_quant_bf16_case10(self):
        shape0 = (352, 1536)
        shape1 = (352)
        input0 = np.random.uniform(low=-5, high=5, size=shape0).astype(np.float32)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0).bfloat16()],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])
        
    @op_test.only_910b
    def test_dynamic_quant_bf16_case11(self):
        shape0 = (1024, 1536)
        shape1 = (1024)
        input0 = np.random.uniform(low=-5, high=5, size=shape0).astype(np.float32)
        PARAM["asymmetric"] = False
        self.set_param(OP_NAME, PARAM)
        self.execute([torch.from_numpy(input0).bfloat16()],
                     [torch.zeros(shape0, dtype=torch.int8),
                      torch.zeros(shape1, dtype=torch.float32),
                      torch.zeros(shape1, dtype=torch.float32)])
        
if __name__ == '__main__':
    unittest.main()